packages/core/lib/commands/migrate/runMigrations.js

```diff
@@ -1,6 +1,110 @@
+
+async function usesConsoleLog(artifactJson) {
+  const debugLog = debug.extend("test");
+  debugLog("Artifact: %o", artifactJson);
+
+  //create a parser to get the value of jsonpath .deployedBytecode
+  const parser = JSONStream.parse(["deployedBytecode"]);
+  const stream = createReadStream(artifactJson).pipe(parser);
+
+  return new Promise((resolve, reject) => {
+    stream.on("data", data => {
+      //JSONParse will emit the entire string/value
+      //so initiate stream cleanup here
+      stream.destroy();
+      const usesConsoleLog = consoleLogRex.test(data);
+      debugLog("usesConsoleLog:", usesConsoleLog);
+      resolve(usesConsoleLog);
+    });
+
+    stream.on("error", err => {
+      stream.destroy();
+      debugLog("onError: %o", err);
+      reject(err);
+    });
+  });
+}
+
+async function findArtifactsThatUseConsoleLog(buildDir) {
+  const debugLog = debug.extend("dirty-files");
+  const files = await readdir(buildDir);
+
+  const artifacts = [];
+  await Promise.allSettled(
+    files.map(async fn => {
+      if (fn.endsWith(".json")) {
+        try {
+          const itLogs = await usesConsoleLog(path.join(buildDir, fn));
+          if (itLogs) {
+            artifacts.push(fn);
+          }
+        } catch (e) {
+          debugLog("Promise failure: %o", e.message);
+        }
+      }
+    })
+  );
+  return artifacts;
+}
 
 module.exports = async function (config) {
+  const debugLog = debug.extend("guard");
+  // only check if deploying on MAINNET
+  // NOTE: this includes Ethereum Classic as well as Ethereum as they're only
+  // distinguishable by checking their chainIds, 2 and 1 respectively.
+  if (config.network_id === 1) {
+    debugLog("solidityLog guard for mainnet");
+    try {
+      const buildDir = config.contracts_build_directory;
+      const loggingArtifacts = await findArtifactsThatUseConsoleLog(buildDir);
+
+      debugLog(`${loggingArtifacts.length} consoleLog artifacts detected`);
+      debugLog(
+        "config.solidityLog.preventConsoleLogMigration: " +
+          config.solidityLog.preventConsoleLogMigration
+      );
+
+      if (loggingArtifacts.length) {
+        console.warn(
+          `${os.EOL}Solidity console.log detected in the following assets:`
+        );
+        console.warn(loggingArtifacts.join(", "));
+        console.warn();
+
+        if (config.solidityLog.preventConsoleLogMigration) {
+          throw new TruffleError(
+            "You are trying to deploy contracts that use console.log." +
+              os.EOL +
+              "Please fix, or disable this check by setting solidityLog.preventConsoleLogMigration to false" +
+              os.EOL
+          );
+        }
+      }
+    } catch (err) {
+      if (err instanceof TruffleError) throw err;
+
+      debugLog("Unexpected error %o:", err);
+      // Something went wrong while inspecting for console log.
+      // Log warning and skip the remaining logic in this branch
```

Comment:
do we want to continue with migrations if the user has this turned on and something goes wrong?

Reply:
Ah, good question. Originally, my thinking was to stop if detecting consoleLog usage, letting the workflow's next steps handle the other exceptions. What do you think?

Reply:
I'm not really sure what we should do. Maybe this is fine for now.

Reply:
This logic tests if console.log code exists in the `bytecode` before migration, and I wanted the errors to reflect that concern. Migration logic still has to kick in and errors are already handled. In the future, we could have a static analysis step during compilation that would persist its outcome, so we query the artifact instead of analyzing on the fly.

Reply:
I'm going to resolve this for now, feel free to reopen if I misunderstood your intents.